go/vt/srvtopo/keyspace_filtering_server.go

```diff
@@ -0,0 +1,105 @@
+/*
+Copyright 2018 The Vitess Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package srvtopo
+
+import (
+	"fmt"
+
+	"golang.org/x/net/context"
+
+	topodatapb "vitess.io/vitess/go/vt/proto/topodata"
+	vschemapb "vitess.io/vitess/go/vt/proto/vschema"
+	"vitess.io/vitess/go/vt/topo"
+)
+
+// ErrNilUnderlyingServer is returned when attempting to create a new keyspace
+// filtering server if a nil underlying server implementation is provided.
+var ErrNilUnderlyingServer = fmt.Errorf("Unable to construct filtering server without an underlying server")
+
+// NewKeyspaceFilteringServer constructs a new server based on the provided
+// implementation that prevents the specified keyspaces from being exposed
+// to consumers of the new Server.
+func NewKeyspaceFilteringServer(underlying Server, selectedKeyspaces []string) (Server, error) {
+	if underlying == nil {
+		return nil, ErrNilUnderlyingServer
+	}
+
+	keyspaces := map[string]bool{}
+	for _, ks := range selectedKeyspaces {
+		keyspaces[ks] = true
+	}
+
+	return keyspaceFilteringServer{
+		server:          underlying,
+		selectKeyspaces: keyspaces,
+	}, nil
+}
+
+type keyspaceFilteringServer struct {
+	server          Server
+	selectKeyspaces map[string]bool
+}
+
+func (ksf keyspaceFilteringServer) GetTopoServer() *topo.Server {
+	return ksf.server.GetTopoServer()
```

Comment:
I actually think we shouldn't return the underlying server here, and instead we should change the `GetTopoServer()` signature to return `(*topo.Server, error)` and return an error of "Cannot modify underlying topology server when keyspace filtering is enabled" here.

The reason for this is that `GetTopoServer()` is only needed as part of the vschema DDL management implementation at vtgate (see https://github.com/vitessio/vitess/blob/master/go/vt/vtgate/vschema_manager.go#L125-L147), and not as part of the general query serving path.

The vschema ddl implementation assumes that the SrvVschema that it is operating on is complete, and as such when vschema DDL operations make changes to the vschema, it writes back the updated SrvVschema to each cell.

When used with keyspace filtering, this would have the very undesirable effect of removing the filtered-out keyspaces from the SrvVschema globally.

So, I think it's only safe to allow vschema DDL operations when keyspace filtering is not in use, and that the best way to enforce that is to hide the underlying topo server altogether here.

Reply:
This seems reasonable and I'll start the process.

Based on [this release doc](https://github.com/vitessio/vitess/blob/master/doc/internal/ReleaseInstructions.md#major-release-vx) I think that this is covered by our external API so we shouldn't need to worry too much about versioning implications so long as the build passes CI?

In any case, assuming this update goes as planned, once this PR lands I'll open an issue that captures the work to patch visible keyspace updates into the global SrvVSchema